ibis/backends/pyspark/compiler.py

```diff
@@ -1,2137 +1,502 @@
+            return self.if_(pos > 0, pos + start, 0)
 
+        return self.f.instr(arg, substr)
 
-@compiles(ops.Substring)
-def compile_substring(t, op, raw: bool = False, **kwargs):
-    src_column = t.translate(op.arg, raw=raw, **kwargs)
-    start = t.translate(op.start, **kwargs, raw=True) + 1
-    length = t.translate(op.length, **kwargs, raw=True)
+    @visit_node.register(ops.RegexReplace)
+    def visit_RegexReplace(self, op, *, arg, pattern, replacement):
+        return self.f.regexp_replace(arg, pattern, replacement)
 
-    if any_of((start, length), pyspark.sql.Column):
-        raise NotImplementedError(
-            "Specifying `start` or `length` with column expressions is not supported."
-        )
+    @visit_node.register(ops.JSONGetItem)
+    def visit_JSONGetItem(self, op, *, arg, index):
+        if op.index.dtype.is_integer():
+            fmt = "$[%s]"
+        else:
+            fmt = "$.%s"
+        path = self.f.format_string(fmt, index)
+        return self.f.get_json_object(arg, path)
+
+    @visit_node.register(Window)
+    def visit_Window(self, op, *, func, group_by, order_by, **kwargs):
+        if isinstance(op.func, ops.Analytic):
+            # spark disallows specifying boundaries for lead/lag
+            if order_by:
+                order = sge.Order(expressions=order_by)
+            else:
+                # pyspark requires an order by clause for lag/lead
+                order = sge.Order(expressions=[sge.NULL])
+            return sge.Window(this=func, partition_by=group_by, order=order)
+        else:
+            return super().visit_node(
+                op, func=func, group_by=group_by, order_by=order_by, **kwargs
+            )
 
-    if start < 0:
-        raise NotImplementedError("`start < 0` is not supported.")
+    @visit_node.register(ops.JoinLink)
+    def visit_JoinLink(self, op, **kwargs):
+        if op.how == "asof":
+            raise com.UnsupportedOperationError(
+                "ASOF joins are not supported by Spark SQL yet and LATERAL joins "
+                "raise an analysis error if the lateral subquery is limited which "
+                "would be necessary to emulate ASOF joins. Once this is fixed "
+                "upstream, we can add support for ASOF joins."
+            )
+        return super().visit_JoinLink(op, **kwargs)
+
+    @visit_node.register(ops.RowID)
+    @visit_node.register(ops.TimestampBucket)
+    def visit_Undefined(self, op, **_):
+        raise com.OperationNotDefinedError(type(op).__name__)
+
+
+_SIMPLE_OPS = {
+    ops.ArrayDistinct: "array_distinct",
+    ops.ArrayFlatten: "flatten",
+    ops.ArrayIntersect: "array_intersect",
+    ops.ArrayRemove: "array_remove",
+    ops.ArraySort: "array_sort",
+    ops.ArrayUnion: "array_union",
+    ops.EndsWith: "endswith",
+    ops.Hash: "hash",
+    ops.Log10: "log10",
+    ops.LPad: "lpad",
+    ops.LStrip: "ltrim",
+    ops.RPad: "rpad",
+    ops.RStrip: "rtrim",
+    ops.StringAscii: "ascii",
```

Comment:
This shouldn't be necessary, it should be in the base sqlglot compiler.

Reply:
Removing.